libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp

```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+void AP_CRSF_Telem::setup_custom_telemetry() {
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+
+    if (_custom_telem_init_done) {
+        return;
+    }
+
+    // we need crossfire firmware version
+    if (_crsf_version.pending) {
+        return; 
+    }
+
+    AP_Frsky_SPort_Passthrough* passthrough = AP::frsky_passthrough_telem();
+    if (passthrough == nullptr) {
+        return;
+    }
+    
+    // setup the frsky scheduler for crossfire
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LAT);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LON);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::TEXT);
+    passthrough->set_scheduler_entry_min_period(AP_Frsky_SPort_Passthrough::ATTITUDE, 120); // 5Hz
+
+    // setup the crossfire scheduler for custom telemetry
+    set_scheduler_entry(BATTERY, 1000, 1000);       // 1Hz
+    set_scheduler_entry(ATTITUDE, 1000, 1000);      // 1Hz
+    set_scheduler_entry(FLIGHT_MODE, 1200, 2000);   // 0.5Hz
+    set_scheduler_entry(HEARTBEAT, 2000, 5000);     // 0.2Hz
+    
+    _telem_rf_mode = get_rf_mode();
+    // setup custom telemetry for current rf_mode
+    update_custom_telemetry_rates(_telem_rf_mode);
+
+    gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: custom telemetry initialized");
+    _custom_telem_init_done = true;
+}
+
+void AP_CRSF_Telem::update_custom_telemetry_rates(uint8_t rf_mode) {
+    if (rf_mode == AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ) {
+        // custom telemetry for high data rates
+        set_scheduler_entry(GPS, 550, 280);           // 3Hz
+        set_scheduler_entry(PASSTHROUGH, 50, 50);     // 20Hz
+        set_scheduler_entry(STATUS_TEXT, 100, 500);   // 2Hz
+    } else {
+        // custom telemetry for low data rates
+        set_scheduler_entry(GPS, 550, 1000);              // 1Hz
+        set_scheduler_entry(PASSTHROUGH, 500, 3000);      // 0.5Hz
+        set_scheduler_entry(STATUS_TEXT, 600, 2000);      // 0.5Hz
+    }
+}
+
+void AP_CRSF_Telem::process_rf_mode_changes()
+{
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+    const uint8_t current_rf_mode = get_rf_mode();
+    if ( _telem_rf_mode != current_rf_mode) {
+        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d", _telem_rf_mode, current_rf_mode);
+        update_custom_telemetry_rates(current_rf_mode);
+        _telem_rf_mode = current_rf_mode;
+    }
+}
+
+// return custom frame id based on fw version
+uint8_t AP_CRSF_Telem::get_custom_telem_frame_id() const 
+{
+    if (!_crsf_version.pending && _crsf_version.major >= 4 && _crsf_version.minor >= 6) {
+        return AP_RCProtocol_CRSF::CRSF_FRAMETYPE_CUSTOM_TELEM;
+    }
+    return AP_RCProtocol_CRSF::CRSF_FRAMETYPE_CUSTOM_TELEM_LEGACY;
+}
+
+uint8_t AP_CRSF_Telem::get_rf_mode() const
+{
+    AP_RCProtocol_CRSF* crsf = AP::crsf();
+    if (crsf == nullptr) {
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ;
```

Comment:
Should there really be a `NONE` entry in that enumeration?

Reply:
don't know really, the radio will switch to whatever it can based on link quality, NONE might mean not detected yet, but what do we do if we did not detect it?,My assumption was I go for 150Hz and eventaully detect it later

Reply:
perhaps better an UNKNOWN, done